### ISSUE TITLE
Issue in loading ParserAnnotator in Android

### ISSUE THREAD

**Issue by jflipts** [stanfordnlp/CoreNLP#861]
I have a working version that runs properly in desktop Java but throws a NullPointerExeception when porting it to Android. I cannot use the webservice because the application should run offline.

This is how I added CoreNLP to gradle
``` 
    implementation('edu.stanford.nlp:stanford-corenlp:3.9.2') {
        exclude group: 'org.apache.lucene'
        exclude group: 'com.sun.xml.bind'
    }
    implementation('edu.stanford.nlp:stanford-corenlp:3.9.2:models') {
        exclude group: 'org.apache.lucene'
        exclude group: 'com.sun.xml.bind'
    }
```
Other info that might be usefull: minSdkVersion = 22, errors occur on Oneplus 6t (Api 28), pixel2 (Api 28), ...

This is how I build the pipeline
```
        Properties props = new Properties();
        props.setProperty("annotators", "tokenize,ssplit,parse,pos,lemma");
        StanfordCoreNLP pipeline = new StanfordCoreNLP(props);
```

But building this pipeline throws following error:
```
2019-04-01 17:10:02.972 27200-27242/com.aurora.aurora E/AndroidRuntime: FATAL EXCEPTION: RxComputationThreadPool-3
    Process: com.aurora.aurora, PID: 27200
    io.reactivex.exceptions.OnErrorNotImplementedException: Attempt to invoke interface method 'int java.util.List.size()' on a null object reference
        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
        at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
        at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:67)
        at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:64)
        at io.reactivex.internal.operators.observable.ObservableFilter$FilterObserver.onNext(ObservableFilter.java:52)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
        at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'int java.util.List.size()' on a null object reference
        at edu.stanford.nlp.util.HashIndex.size(HashIndex.java:94)
        at edu.stanford.nlp.parser.lexparser.BinaryGrammar.init(BinaryGrammar.java:228)
        at edu.stanford.nlp.parser.lexparser.BinaryGrammar.readObject(BinaryGrammar.java:215)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1067)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1902)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1803)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2002)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1926)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1803)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1353)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:373)
        at edu.stanford.nlp.io.IOUtils.readObjectFromURLOrClasspathOrFileSystem(IOUtils.java:309)
        at edu.stanford.nlp.parser.common.ParserGrammar.loadModel(ParserGrammar.java:184)
        at edu.stanford.nlp.pipeline.ParserAnnotator.loadModel(ParserAnnotator.java:219)
        at edu.stanford.nlp.pipeline.ParserAnnotator.<init>(ParserAnnotator.java:121)
        at edu.stanford.nlp.pipeline.AnnotatorImplementations.parse(AnnotatorImplementations.java:113)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.lambda$getNamedAnnotators$11(StanfordCoreNLP.java:529)
        at edu.stanford.nlp.pipeline.-$$Lambda$StanfordCoreNLP$IYXSs0FaR-yYo5NN51RbYarn0xQ.apply(Unknown Source:4)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.lambda$null$30(StanfordCoreNLP.java:602)
        at edu.stanford.nlp.pipeline.-$$Lambda$StanfordCoreNLP$bGaSmS8dhCkVuNcmkJjJlAI05AU.get(Unknown Source:6)
        at edu.stanford.nlp.util.Lazy$3.compute(Lazy.java:126)
        at edu.stanford.nlp.util.Lazy.get(Lazy.java:31)
        at edu.stanford.nlp.pipeline.AnnotatorPool.get(AnnotatorPool.java:149)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:251)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:192)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:188)
2019-04-01 17:10:02.972 27200-27242/com.aurora.aurora E/AndroidRuntime:     at com.aurora.internalservice.internalnlp.InternalNLP.buildPipeline(InternalNLP.java:51)
        at com.aurora.internalservice.internalnlp.InternalNLP.<init>(InternalNLP.java:30)
        at com.aurora.kernel.PluginInternalServiceCommunicator.processFileWithInternalProcessor(PluginInternalServiceCommunicator.java:107)
        at com.aurora.kernel.PluginInternalServiceCommunicator.lambda$new$0(PluginInternalServiceCommunicator.java:40)
        at com.aurora.kernel.-$$Lambda$PluginInternalServiceCommunicator$6apnywtdZcS_QxLe7FHbnfTTEWI.accept(Unknown Source:4)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        	... 11 more
```

**Comment by J38** [comment]
Do you have the models jar accessible in your Android app?  If you look at the stack trace you can see the failure involves trying to read a resource, which is probably missing.  I am unclear if it's really possible to integrate Stanford CoreNLP into Android apps because of size restrictions.  You need the `stanford-corenlp-3.9.2-models.jar` to be accessible though.

**Comment by erksch** [comment]
@jflipts For me, this error occurs when using ner, did you manage to fix it?

**Comment by jflipts** [comment]
@erksch 
We ended up rewritting our service to only support TokenizerAnnotator, WordsToSentencesAnnotator and POSTaggerAnnotator. As these were the only ones we actually got working within the heavy memory constraints of Android. If I would do it again I wouldn't try CoreNLP on Android anymore as it is a big headache and just try to use the webservice instead. This will make your app way smaller as the models are quite big and also performance will be much better.

Another thing to note is that CoreNLP will only work on Android 8 or higher due to other issues, so the userbase for the app will be quite small either way.

**Comment by erksch** [comment]
@jflipts 
Thank you for your fast reply. That's pretty sad to hear, but I have some hope left because we are building for B2B where we can choose the devices that our customers will use, thus the Android version, too. But at the same time, offline capability is one of our requirements.

You are right, it's a big headache and there are a lots of problems but that some annotators actually do work is kind of a win. I am thinking about writing an article about getting CoreNLP working on Android (as best as possible) to ease the pain for others.

**Comment by jflipts** [comment]
@erksch 
You can find an example project to get the 3 annotators working on my Github.
https://github.com/jflipts/CoreNLPAndroid

I wish you good luck and proper guide would be really helpful for others.

**Comment by erksch** [comment]
Hey just to let you know, after debugging through the CoreNLP code I found out that some model files can not be loaded properly while others not. The `DEFAULT_PARSER_MODEL = "edu/stanford/nlp/models/lexparser/englishPCFG.ser.gz"` seems to be affected as well. 
Because the model is not loaded properly some properties are null resulting in the error above.

For NER, the code loads the following three models per default (see `NERCombinerAnnotator.java:90`):
- `DEFAULT_NER_THREECLASS_MODEL`
- `DEFAULT_NER_MUC_MODEL`
- `DEFAULT_NER_CONLL_MODEL`

While debugging I found out that the first model is loaded correctly and the second is not, resulting in the `'int java.util.List.size()' on a null object reference` error. 

So I tried applying the first model **only**:
```
props.setProperty("ner.model", DefaultPaths.DEFAULT_NER_THREECLASS_MODEL)
```
And it actually works!

@J38 Maybe a contributor can explain why some models can be loaded and some not and why this could be due to the android environment?
